src/librustc_mir/borrow_check/error_reporting.rs

```diff
@@ -552,9 +552,17 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
                 // created by `StorageDead` and at the beginning
                 // of a function.
             } else {
+                let mut mpis = vec![mpi];
+
+                let mut curr_mpi = mpi;
+                while let Some(parent_mpi) = &self.move_data.move_paths[curr_mpi].parent {
```

Comment:
you probably don't need the `&` here or the `*` on the lines below

Reply:
I'd also like to see a comment explaining this. Something like:

---

If we are found a use of `a.b.c` which was in error, then we want to look for moves not only of `a.b.c` but also `a.b` and `a`.

Reply:
Writing that out makes me realize that probably there is a similar bug the other way... e.g., what happens here?

```rust
fn main() {
    let x = (vec![1, 2, 3], );
    drop(x.0);
    drop(x);
}
```

Ah, trying it on play, it seems to work ok. Interesting. I guess that the `moves` data already handles this case -- i.e., a move of `x.0` is marked as a move of `x`. Maybe add this to the comment:

---

Note that the moves data already includes "parent" paths, so we don't have to worry about the other case: that is, if there is a move of `a.b.c`, it is already marked as a move of `a.b` and `a` as well, so we will generate the correct errors there. 

---

but @pnkfelix can you verify that comment? ☝️

Reply:
Also, I wonder if we want to make a helper for iterating over all parents...

Reply:
@nikomatsakis everything is fixed.